crates/holochain_state/src/mutations.rs

```diff
@@ -907,6 +906,40 @@ pub fn schedule_fn(
     Ok(())
 }
 
+/// Force remove a countersigning session from the source chain.
+///
+/// This is a dangerous operation and should only be used:
+/// - If the countersigning workflow has determined to a reasonable level of confidence that other
+///   peers abandoned the session.
+/// - If the user decides to force remove the session from their source chain when the
+///   countersigning session is unable to make a decision.
+///
+/// Note that this mutation is defensive about sessions that have any of their ops published to the
+/// network. If any of the ops have been published, the session cannot be removed.
+pub fn remove_countersigning_session(
+    txn: &mut Transaction,
+    cs_action: Action,
+    cs_entry_hash: EntryHash,
+) -> StateMutationResult<()> {
+    // Check, just for paranoia's sake that the countersigning session is not fully published.
+    // It is acceptable to delete a countersigning session that has been written to the source chain,
+    // with signatures published. As soon as the session's ops have been published to the network,
```

Comment:
@ThetaSinner there's a spreadsheet that i made trying to cover all these possible states, it's pinned in the mattermost channel

might be worth reviewing that

also when you say for 1. that you're "up waiting for signatures", it might be worth checking with the AAA to see if they think the session completed before removing it from your own chain

there could be some situation where all the signatures made it to someone else, but you never got it yourself

all of this is just probabalistic anyway, there's no way to truly 100% guarantee a safe deletion, so adding in more guards over time could be followups to make it safer

Reply:
This has been done, a timeout will now result in at least one (configurable for more) round of checking with AAAs for the session. Network errors are treated as a failure to check, so if you have network issues Holochain will loop until it does at least one successful query to a group of AAAs to find out what happened to the session.